Explorer/Assets/DCL/Interaction/PlayerOriginated/Systems/ProcessPointerEventsSystem.cs

```diff
@@ -0,0 +1,112 @@
+using Arch.Core;
+using Arch.System;
+using Arch.SystemGroups;
+using Arch.SystemGroups.DefaultSystemGroups;
+using DCL.ECSComponents;
+using DCL.Interaction.PlayerOriginated.Components;
+using DCL.Interaction.PlayerOriginated.Utility;
+using DCL.Interaction.Utility;
+using Diagnostics.ReportsHandling;
+using ECS.Abstract;
+using System.Collections.Generic;
+
+namespace DCL.Interaction.PlayerOriginated.Systems
+{
+    [UpdateInGroup(typeof(PresentationSystemGroup))]
+    [UpdateAfter(typeof(PlayerOriginatedRaycastSystem))]
+    [LogCategory(ReportCategory.INPUT)]
+    public partial class ProcessPointerEventsSystem : BaseUnityLoopSystem
+    {
+        private readonly IEntityCollidersGlobalCache entityCollidersGlobalCache;
+        private readonly IReadOnlyDictionary<InputAction, UnityEngine.InputSystem.InputAction> sdkInputActionsMap;
+
+        internal ProcessPointerEventsSystem(World world,
+            IReadOnlyDictionary<InputAction, UnityEngine.InputSystem.InputAction> sdkInputActionsMap,
+            IEntityCollidersGlobalCache entityCollidersGlobalCache) : base(world)
+        {
+            this.sdkInputActionsMap = sdkInputActionsMap;
+            this.entityCollidersGlobalCache = entityCollidersGlobalCache;
+        }
+
+        protected override void Update(float t)
+        {
+            ProcessRaycastResultQuery(World);
+        }
+
+        private bool TryGetPreviousEntityInfo(in HoverStateComponent stateComponent, out GlobalColliderEntityInfo globalColliderEntityInfo)
+        {
+            if (!stateComponent.LastHitCollider) // collider was destroyed, nothing to do
+            {
+                globalColliderEntityInfo = default(GlobalColliderEntityInfo);
+                return false;
+            }
+
+            return entityCollidersGlobalCache.TryGetEntity(stateComponent.LastHitCollider, out globalColliderEntityInfo); // scene was destroyed, collider was returned to the pool, nothing to do
+        }
+
+        [Query]
```

Comment:
this method is huge and contains several nesting levels. It leads to high cognitive complexity. I suggest to split it into readable sub-functions

Reply:
Do you have ideas of what we can move to a separate method to improve readability **considerably**?

Reply:
Here an example of extraction to more high-level functions, so reader can skip if he is not needed to dive into low-level. Naming of the functions could be probably improved. Also, I am not sure auto-extraction properly used `ref` in all cases
```

[Query]
        private void ProcessRaycastResult(ref PlayerOriginRaycastResult raycastResult, ref HoverFeedbackComponent hoverFeedbackComponent,
            ref HoverStateComponent hoverStateComponent)
        {
            // Process all PBPointerEvents components to see if any of them is qualified
            hoverFeedbackComponent.Tooltips.Clear();

            bool candidateForHoverLeaveIsValid = TryGetPreviousEntityInfo(in hoverStateComponent, out GlobalColliderEntityInfo previousEntityInfo);
            hoverStateComponent.LastHitCollider = null;

            if (raycastResult.IsValidHit)
            {
                var entityInfo = raycastResult.EntityInfo.Value;

                // External world access should be always synchronized (Global World calls into Scene World)
                using (entityInfo.EcsExecutor.Sync.GetScope())
                {
                    candidateForHoverLeaveIsValid = 
                        ProcessGlobalToScene(raycastResult, ref hoverFeedbackComponent, ref hoverStateComponent, entityInfo.ColliderEntityInfo.EntityReference, 
                            entityInfo.EcsExecutor.World, candidateForHoverLeaveIsValid, previousEntityInfo);
                }
            }

            if (candidateForHoverLeaveIsValid)
                HoverFeedbackUtils.TryIssueLeaveHoverEventForPreviousEntity(in raycastResult, in previousEntityInfo);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private bool ProcessGlobalToScene(PlayerOriginRaycastResult raycastResult,
            ref HoverFeedbackComponent hoverFeedbackComponent, ref HoverStateComponent hoverStateComponent,
            EntityReference entityRef, World world, bool candidateForHoverLeaveIsValid,
            GlobalColliderEntityInfo previousEntityInfo)
        {
            // Entity should be alive and contain PBPointerEvents component to be qualified
            if (!entityRef.IsAlive(world) || !world.TryGet(entityRef, out PBPointerEvents pbPointerEvents))
                return candidateForHoverLeaveIsValid;
            
            hoverStateComponent.LastHitCollider = raycastResult.UnityRaycastHit.collider;

            var newEntityWasHovered = !candidateForHoverLeaveIsValid
                                      || (previousEntityInfo.EcsExecutor.World != world &&
                                          previousEntityInfo.ColliderEntityInfo.EntityReference != entityRef);

            var anyInputInfo = sdkInputActionsMap.Values.GatherAnyInputInfo();
            pbPointerEvents.AppendPointerEventResultsIntent.Initialize(raycastResult.UnityRaycastHit, raycastResult.OriginRay);

            for (var i = 0; i < pbPointerEvents.PointerEvents.Count; i++)
                TryAppendPointerInfo(raycastResult, ref hoverFeedbackComponent, i, pbPointerEvents, anyInputInfo, newEntityWasHovered);

            // Signal to stop issuing hover leave event for the previous entity as it's equal to the current one
            if (candidateForHoverLeaveIsValid && !newEntityWasHovered)
                candidateForHoverLeaveIsValid = false;

            return candidateForHoverLeaveIsValid;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private void TryAppendPointerInfo(PlayerOriginRaycastResult raycastResult,
            ref HoverFeedbackComponent hoverFeedbackComponent, int i, PBPointerEvents pbPointerEvents,
            InteractionInputUtils.AnyInputInfo anyInputInfo, bool newEntityWasHovered)
        {
            var pointerEvent = pbPointerEvents.PointerEvents[i];
            var info = pointerEvent.EventInfo;

            info.PrepareDefaultValues();

            if (!InteractionInputUtils.IsQualifiedByDistance(raycastResult, info)) return;

            // Check Input for validity
            InteractionInputUtils.TryAppendButtonLikeInput(sdkInputActionsMap, pointerEvent, i,
                ref pbPointerEvents.AppendPointerEventResultsIntent, anyInputInfo);

            // Try Append Hover Input
            if (newEntityWasHovered)
                InteractionInputUtils.TryAppendHoverInput(ref pbPointerEvents.AppendPointerEventResultsIntent,
                    PointerEventType.PetHoverEnter, pointerEvent, i);

            // Try Append Hover Feedback
            HoverFeedbackUtils.TryAppendHoverFeedback(sdkInputActionsMap, pointerEvent,
                ref hoverFeedbackComponent, anyInputInfo.AnyButtonIsPressed);
        }
```